lua/autorun/nl_library.lua

```diff
@@ -0,0 +1,609 @@
+				if k == ent.nllplayer:GetRole() or k == ent.nllplayer:GetSubRole() then
+					BlacklistResult = v
+				end
+			end
+		end
+		if IllegalCovertKill[ply:GetSubRole()] then
+			for k,v in pairs(IllegalCovertKill[ply:GetSubRole()]) do -- Roles go next, for special interactions
+				if k == ent.nllplayer:GetSubRole() or k == ent.nllplayer:GetSubRole() or k == ROLE_ALLCALL then
+					BlacklistResult = v -- the significance of subrole going next means you can display specific messages
+				end
+			end
+		end
+		if BlacklistResult ~= "This kill is actually allowed" then
+			if not ply.lastCovertAttempt then ply.lastCovertAttempt = 0 end
+			if ply.lastCovertAttempt + 0.25 < SysTime() then
+				ply:ChatPrint( BlacklistResult ) -- an informative message :^)
+				ply.lastCovertAttempt = SysTime()
+			end
+			return -- no
+		end
+	end
+        plytarg = ent.nllplayer
+        NLL.UnRagdoll(plytarg)
+	plytarg.nllismuted = false
+        if timer.Exists("NLLUnragdoll".. plytarg:UserID()) then
+            timer.Remove( "NLLUnragdoll".. plytarg:UserID())
+        end
+        
+	if ply:IsWalking() then -- alt + e to kill instead of waking
+	    if ply:GetInfoNum("ttt_nll_covertbuddha", 0) == 1 then
+		plytarg:SetHealth(1)
+		return
+	    end
+            local dmgtable = DamageInfo()
+            dmgtable:SetDamage((plytarg:Health())*1.5)
+            dmgtable:SetAttacker(ply)
+            dmgtable:SetDamageForce(Vector(0,0,-1))
+            dmgtable:SetDamagePosition(plytarg:GetPos())
+            dmgtable:SetDamageType(4)
+            dmgtable:SetInflictor(ply)
+            timer.Create("NLLKillTarget".. plytarg:UserID(), 0.2, 1, function()
+                plytarg:TakeDamageInfo( dmgtable )
+            end)
+        end
+    end
+end)
+
+function NLL.PlayerZZZ( ply, pushdir, stuntype, timebefore, timedur )
+	local id = ply:UserID()
+        
+	if stuntype == "ZZZ" or stuntype == "KSR" then -- Kissers cannot be woke
+		if timebefore < 5 then ply.nllismuted = true end -- Instantly mute a player if they happen to get struck by a 'critical tranq'
+		if timer.Exists("NLLSleepIn"..id) then
+			if timer.TimeLeft( "NLLSleepIn"..id ) < 3 then return end
+			local newdelay = math.min( timebefore/3, timer.TimeLeft( "NLLSleepIn"..id )/2 )
+			timer.Remove("NLLSleepIn"..id)
+			timer.Create("NLLSleepIn"..id, newdelay, 1, function()
+				NLL.Ragdoll(ply, pushdir,stuntype)
+				ply.nllismuted = true
+				timer.Create("NLLUnragdoll"..id, timedur, 1, function()
+					NLL.UnRagdoll(ply)
+					ply.nllismuted = false
+				end)
+			end)
+			
+			ply.DrowsyDuration = newdelay
+			net.Start("nllinfopayload")
+				net.WriteString("drowsy")
+				net.WriteTable({["drowsy"] = newdelay})
+			net.Send(ply)
+			return
+		end
+		timer.Create("NLLSleepIn"..id, timebefore, 1, function()
+			NLL.Ragdoll(ply, pushdir,stuntype)
+			ply.nllismuted = true
+			timer.Create("NLLUnragdoll"..id, timedur, 1, function()
+				NLL.UnRagdoll(ply)
+				ply.nllismuted = false
+			end)
+		end)
```

Comment:
Unclean, `NLLSleepIn` uses the same code twice. This should be a localized function to avoid redundancy.